brave/brave-core
# Port SOCKS5 authentication from muon.

This ports the patches, makes them work, and adds unit tests, in separate commits.  (If the unit tests should go again, they can be removed -- much easier this time since all the patches are in different files.)

fix https://github.com/brave/brave-browser/issues/358

## Submitter Checklist:


## Test Plan:
After a build, say a Debug build:
- Build and run the unit tests:
  1. `ninja -C src/out/Debug net:net_unittests`
  2. `./src/out/Debug/net_unittests --gtest_filter='ProxyServerTest.*:SOCKS5ClientSocketTest.*'`
- Manually confirm the behaviour with Tor:
  1. Launch a tor process listening on (say) 9250, e.g. by starting browser-laptop and opening a private tab with Tor.
  2. Launch two independent Brave instances with different proxy server usernames and passwords, e.g.:
     - `./src/out/Debug/brave --user-data-dir=/tmp/test1 --proxy-server=socks5://foo:bar@127.0.0.1:9250 &`
     - `./src/out/Debug/brave --user-data-dir=/tmp/test2 --proxy-server=socks5://baz:quux@127.0.0.1:9250 &`
  3. Load https://check.torproject.org in the two browsers and confirm that they have different IP addresses.

## Reviewer Checklist:
